### PR TITLE
Correct fragile windows CMake assumptions

### PR DESCRIPTION
**Pull request by wrowe**
- Locking CMake to 3.16 breaks all features and corrections applied to
  CMake 3.17 and later, including the correction of the poorly designed
  and now abandoned Windows CRT election policy CMP0091 (see final para
  of the policy description here:
  https://cmake.org/cmake/help/v3.18/policy/CMP0091.html). Locking to
  rev 3.16 from ensures a more difficult transition to CMake-current

- Windows curl builds previously only adjusted the Release and Debug
  builds, and combined with CMP0091 to break other flavors. Update any
  /MD* flags with /MT* present in the base and four alternate build
  flavors, without introducing conflicting flag values or introducing
  a CRT election where one is not present

- Windows clang-cl builds of curl static libs are broken when using
  link-lld.exe because curl appended the dynamic run time flags to the
  static library lib.exe options. While these were ignored/no-op on
  Windows link.exe, they cause link-lld from LLVM/clang-cl compile
  toolchain to fail to parse the library command.

Summary/example of the problems caused by these CMakeLists.txt
elements exists in this bazel-specific bug report;
https://github.com/bazelbuild/rules_foreign_cc/issues/426

### CODE REVIEW

**Comment by jzakrzewski** [comment]
I don't really like the change of the `cmake_minimum_required...`. I understand the problem from the other hand not specifying MAX version causes all policy settings introduced after MIN version to be set to OLD. Setting them to NEW has been one of the points of @DerDakon's changes.
Additionally as far as I can understand the docu correctly, MAX version does not lock anything - it just sets all polices introduced in version up to MAX to NEW.

**Comment by DerDakon** [comment]
Right, it just sets the policies. It tells CMake that this code is only using NEW policies for anything added before and in 3.16, and OLD for everything that comes later. If you test that it works fine with newer CMake versions you may as well bump that number, or if you find that you need only a specific policy introduced later you can do
```cmake
if(POLICY CMP1234)
  cmake_policy(SET CMP1234 NEW)
endif()
```
If CMP0091 is actually causing trouble you can reset it the same way to OLD right after the ```cmake_minumum_version```. The whole point of the ```3.2...3.16``` is that CMake will not warn about any policies introduced in that range that would make a difference, but just select their new version, and doing this while still allow someone to use an older version.

**Comment by wrowe** [comment]
So if we look at CMP0091, it explicitly spells out that 3.18.4 is treating
it as OLD, backing out the broken behavior we observe with the ...3.16
ruleset in 3.18.

**Comment by jzakrzewski** [comment]
Well then, the solution seems to be either:
```
if(POLICY CMP0091)
  cmake_policy(SET CMP0091 OLD)
endif()
```

or (and?) simply 
`cmake_minimum_required(VERSION 3.2..3.18 FATAL_ERROR)` isn't it?

> So if we look at CMP0091, it explicitly spells out that 3.18.4 is treating it as OLD, backing out the broken behavior we observe with the ...3.16 ruleset in 3.18.

BTW: docu for 3.15.7 says the same...

**Comment by DerDakon** [comment]
The documentation says it is treated as old if not set. I _is_ set by the ```cmake_minimum_required``` call to new for all versions of CMake that support the ...-syntax, i.e. 3.12 onward. So every version of CMake that has CMP0091 will treat it as NEW.

**Comment by jzakrzewski** [comment]
Thanks for clarifying, @DerDakon.
This means the right fix here seems to be to explicitly set this policy to OLD.

**Comment by wrowe** [comment]
This would be sensible, but is not in fact what happens...

Setting max as ...3.18 or ...3.18.4 toggles the policy on, just as ...3.16
does.

Setting no max (simply '3.2') defaults the policy to old.

Setting the policy to OLD in versions prior to 3.15 breaks cmake
(unrecogzied policy).

I think the best answer is to force the policy to old to avoid this mess.
Considering the interactions between the many curl dependencies and curl
cmake build, synchronizing the shift would be unwise. If the desire is to
use the new policy and override the variable, a curl-specific toggle is the
wrong way to go about it.

I think we can constrain forcing the old policy only for windows for cmake
3.15+ for curl-forced-libcmt, I'm testing that next.

**Comment by DerDakon** [comment]
You don't need to add extra checks around that policy, as it's only there for 3.15+ and affects only Windows. Just check if it exists as I have written above. The version in ```cmake_minimum_required``` doesn't need to be changed, and I would advise doing so only if explicitely checking with a newer version and reviewing the new policies.

Given that this change has gone undetected before, maybe add a dedicated test build covering this, which would help when later porting that to CMP0091 NEW to see if everything is correct.

**Comment by wrowe** [comment]
Thanks for all the clarifications @jzakrzewski and @DerDakon, I think I follow the logic of max now, and I agree with not changing it while we can revert one specific policy. Revised PR pushed.

The exact scenario we use is that we don't target _RELEASE or _DEBUG, but a custom target which deliberately sets no model-specific compile or link flags. These are propagated to cmake directly by bazel's rules_foreign_cc logic, which injects the entire CMAKE_C_FLAGS, CMAKE_LINK_FLAGS etc as complete fields for the targeted build model. It becomes impossible to bypass the CMP0091 logic, and that logic isn't used by any other dependency of curl which are similarly cmake builds.

I'm happy to help look at what tests might best verify all four windows builds plus this more unusual case, although I'm a bit
stretched for time ahead of another project beta release.

**Comment by DerDakon** [comment]
I would not call it a "CMake bug" until it's clear that there is really a bug. Please open a CMake issue describing what problems you have, maybe the CMake guys with Windows experience can find a better way to sort that out.

**Comment by wrowe** [comment]
> Please open a CMake issue describing what problems you have, maybe the CMake guys with Windows experience can find a better way to sort that out.

That's a fair comment, and I will be surfacing a number of CMake proposals related to windows. (E.g. RelWithDebInfo is currently linking incrementally, never a good thing.) The 10k foot problem we encounter with CMake is a lack of flexibility in overriding a project or cmake choices without resorting to patching CMakeLists.txt, which many of my cohorts find problematic.

Unless the new CMP0091 compile and link overrides can themselves be overridden, this is a regression and nonstarter in our use cases.

File: CMakeLists.txt [hunk 1]
```diff
@@ -39,6 +39,11 @@
 # (From Daniel Stenberg) The gcc command line use neither -g nor any -O options. As a developer, I also treasure our configure scripts's --enable-debug option that sets a long range of "picky" compiler options.
 cmake_minimum_required(VERSION 3.2...3.16 FATAL_ERROR)
 
+# Revert CMake bug triggered by curl's defined max CMake policy version, see https://gitlab.kitware.com/cmake/cmake/-/issues/21288
```
Comment (jzakrzewski) [inline CMakeLists.txt:42]: I don't think calling it a CMake bug is right - it's problematic, sure but very much intentional behaviour on the CMake side.
Just use this comment to explain what and why is being set here so next person does not have to scratch their head looking what the CMP0091 is.

File: CMakeLists.txt [hunk 2]
```diff
@@ -246,9 +251,12 @@ if(ENABLE_MANUAL)
 endif()
 
 if(CURL_STATIC_CRT)
-  set(CMAKE_MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")
```
Comment (jzakrzewski) [inline CMakeLists.txt:249]: Why removing this line?

**Comment by bagder** [comment]
any news on this work?

**Comment by bagder** [comment]
This PR has not gotten sufficient attention for a long while with outstanding questions and remarks. It  will therefore be closed. This is not rejecting the idea nor the effort here, just that it isn't ready to be merged and it does not seem to be moving in that direction either.

**Comment by wrowe** [comment]
Relocated the PR itself to https://github.com/wrowe/curl/tree/fix-cmake-windows-targets for future discussion when I can come back to this. Short term we have patched around this for use with bazel. In our cmake use case, we inject all CFLAGS / LDFLAGS directly as all the rest of the components are compiled, so manipulating these when the target is 'Bazel' and is not one of the corresponding Debug/Release/RelWithDebugInfo options remains a problem. Will open a fresh PR, more clearly described and summary of this PR when I have free cycles.

**Pull request closed without merging**